x-pack/plugins/security_solution/public/entity_analytics/components/asset_criticality_file_uploader/components/result_step.tsx

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import {
+  EuiButtonEmpty,
+  EuiCallOut,
+  EuiCodeBlock,
+  EuiFlexGroup,
+  EuiHorizontalRule,
+  EuiSpacer,
+  useEuiTheme,
+} from '@elastic/eui';
+import React from 'react';
+import { FormattedMessage } from '@kbn/i18n-react';
+import { i18n } from '@kbn/i18n';
+import { css } from '@emotion/react';
+import type { AssetCriticalityCsvUploadResponse } from '../../../../../common/entity_analytics/asset_criticality/types';
+
+export const AssetCriticalityResultStep: React.FC<{
+  result?: AssetCriticalityCsvUploadResponse;
+  validLinesAsText: string;
+  errorMessage?: string;
+  onReturn: () => void;
+}> = React.memo(({ result, validLinesAsText, errorMessage, onReturn }) => {
+  const { euiTheme } = useEuiTheme();
+
+  if (result === undefined || errorMessage !== undefined) {
+    return (
+      <>
+        <EuiCallOut
+          data-test-subj="asset-criticality-result-step-error"
+          title={
+            <FormattedMessage
+              defaultMessage="Asset criticality assignment failed."
+              id="xpack.securitySolution.entityAnalytics.assetCriticalityResultStep.errorMessage"
+            />
+          }
+          color="danger"
+          iconType="error"
+        >
+          {errorMessage && errorMessage}
+        </EuiCallOut>
+        <ResultStepFooter onReturn={onReturn} />
+      </>
+    );
+  }
+
+  if (result.stats.failed === 0) {
+    return (
+      <>
+        <EuiCallOut
+          data-test-subj="asset-criticality-result-step-success"
+          title={i18n.translate(
+            'xpack.securitySolution.entityAnalytics.assetCriticalityResultStep.successTitle',
+            { defaultMessage: 'Success' }
+          )}
+          color="success"
+          iconType="checkInCircleFilled"
+        >
+          <FormattedMessage
+            defaultMessage="Your asset criticality assignments have been successfully mapped."
+            id="xpack.securitySolution.entityAnalytics.assetCriticalityResultStep.successMessage"
+          />
+        </EuiCallOut>
+        <ResultStepFooter onReturn={onReturn} />
+      </>
+    );
+  }
+
+  const annotations = result.errors.reduce<Record<number, string>>((acc, e) => {
```

Comment:
this feels a bit unclear. I think we can use `lodash` `sortBy` followed by a `map` to achieve the same result in a more explainable/readable manner

Reply:
`Errors` is an array, and `Annotations` is an object. So, we need to reduce an array into an object.
I understand that `Array.reduce` can be tricky to read. So I transformed it into a `forEach`.

```ts
  const annotations: Record<number, string> = {};

  errors.forEach((e) => {
    annotations[e.index] = e.message;
  });

  return annotations;
```

I also moved it to a helper and add a unit test for documentation.